File: pi/trajectory_visualizer.py
"""
Shot Trajectory Visualizer  —  Team 539
========================================
Draws a 2D side-view of the current physics-calculated shot and streams it as
an MJPEG feed.  Add the following URL as a camera widget in Elastic:

    http://10.5.39.XX:1184/stream.mjpg

Canvas layout (side view, x = downrange distance, y = height above floor)
--------------------------------------------------------------------------
  Left:    Robot box (grey outline) with green shooter-exit dot + launch arrow
  Right:   Hub / target box (white outline) with scoring opening
  Curve:   Simulated parabolic trajectory line (white)
  Ball:    Yellow dot at the hub scoring opening
  Overlay: Distance, flywheel speed, angle, efficiency, drag coefficient

Runs at 10 Hz — the trajectory is static (no animation), so high FPS is
unnecessary and this conserves Pi CPU for the other co-processors.

NetworkTables read
------------------
  UltraShooter/Avg Distance to Hub (m)            — distance to hub
  UltraShooter/Target ft/s                        — current flywheel setpoint
  UltraShooter/Physics Velocity ft/s              — physics-calculated speed
  UltraShooter/Launch Angle (deg)                 — fixed launch angle
  UltraShooter/Hood Height From Floor (in)        — hood exit height
  UltraShooter/Hub Center Height From Floor (in)  — hub center height
  UltraShooter/Pi Active                          — whether Pi physics is live
  ShooterTuner/Params/FlywheelEfficiency          — efficiency (0–1)
  ShooterTuner/Params/DragCoefficient             — drag constant (kg/m)
  ShooterTuner/Params/BallMassLbs                 — ball mass (lbs)
"""

import math
import time
import threading

import cv2
import numpy as np
import ntcore
from http.server import HTTPServer, BaseHTTPRequestHandler

# ── Stream settings ────────────────────────────────────────────────────────────

STREAM_PORT = 1184
CANVAS_W    = 800
CANVAS_H    = 420
FPS         = 10      # static diagram — low FPS sufficient

# ── Colours (BGR) ─────────────────────────────────────────────────────────────

_BG          = (28,  28,  28)
_WHITE       = (255, 255, 255)
_GREY        = (140, 140, 140)
_DARK_GREY   = (70,  70,  70)
_GREEN       = (80,  200,  80)
_YELLOW      = (0,   215, 255)   # bright yellow in BGR
_TRAJ_COLOR  = (210, 210, 210)
_TEXT_COLOR  = (225, 225, 225)
_LABEL_COLOR = (180, 200, 255)
_DIM_RED     = (80,   80, 200)

# ── Shared frame buffer (producer: render loop / consumer: HTTP handlers) ─────

_frame_lock  = threading.Lock()
_latest_jpeg = b""


# ── MJPEG HTTP server ──────────────────────────────────────────────────────────

class _MjpegHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path != "/stream.mjpg":
            self.send_error(404)
            return
        self.send_response(200)
        self.send_header("Content-Type",
                         "multipart/x-mixed-replace; boundary=frame")
        self.end_headers()
        try:
            while True:
                with _frame_lock:
                    jpeg = _latest_jpeg
                if jpeg:
                    self.wfile.write(b"--frame\r\nContent-Type: image/jpeg\r\n\r\n")
                    self.wfile.write(jpeg)
                    self.wfile.write(b"\r\n")
                time.sleep(1.0 / FPS)
        except (BrokenPipeError, ConnectionResetError):
            pass

    def log_message(self, *_):
        pass   # suppress access-log noise


def _start_server():
    HTTPServer(("", STREAM_PORT), _MjpegHandler).serve_forever()


# ── Physics helpers ────────────────────────────────────────────────────────────

def _simulate_trajectory(v0_mps: float, angle_rad: float,
                          drag_per_mass: float, hood_height_m: float,
                          dt: float = 0.005) -> list[tuple[float, float]]:
    """
    Euler-integrates the trajectory.  Returns a list of (x, y) world-space
    coordinates in metres, with y measured from the floor (not the hood exit).
    Stops when the ball hits the floor or travels > 12 m.
    """
    vx = v0_mps * math.cos(angle_rad)
    vy = v0_mps * math.sin(angle_rad)
    x  = 0.0
    y  = hood_height_m   # start at hood exit above the floor
    pts: list[tuple[float, float]] = [(x, y)]
    for _ in range(4000):
        speed = math.sqrt(vx * vx + vy * vy)
        vx += (-drag_per_mass * speed * vx) * dt
        vy += (-9.81 - drag_per_mass * speed * vy) * dt
        x  += vx * dt
        y  += vy * dt
        pts.append((x, y))
        if y < -0.1 or x > 12.0:
            break
    return pts


# ── Canvas rendering ──────────────────────────────────────────────────────────

def _draw_frame(distance_m: float, v_flywheel_fps: float, angle_deg: float,
                hood_h_m: float, hub_h_m: float,
                efficiency: float, drag_coeff: float, ball_mass_lbs: float,
                pi_active: bool) -> bytes:
    """
    Renders one trajectory frame and returns JPEG bytes.
    """
    angle_rad     = math.radians(angle_deg)
    ball_mass_kg  = ball_mass_lbs * 0.453592
    drag_per_mass = drag_coeff / max(ball_mass_kg, 0.001) if drag_coeff > 0 else 0.0
    v0_ball_mps   = (v_flywheel_fps / 3.28084) * efficiency

    # ── Compute trajectory ────────────────────────────────────────────────────
    traj: list[tuple[float, float]] = []
    if distance_m > 0.1 and v0_ball_mps > 0.5:
        traj = _simulate_trajectory(v0_ball_mps, angle_rad, drag_per_mass, hood_h_m)

    # ── World bounding box for the canvas ────────────────────────────────────
    x_max = max(distance_m * 1.12, 1.5)
    y_max = hub_h_m * 1.45
    if traj:
        y_max = max(y_max, max(p[1] for p in traj) * 1.18)

    # ── Layout margins ────────────────────────────────────────────────────────
    ML = 58   # left  margin (space for y-axis labels)
    MR = 18   # right margin
    MT = 32   # top   margin (title)
    MB = 38   # bottom margin (x-axis labels)
    pw = CANVAS_W - ML - MR   # plot width  (pixels)
    ph = CANVAS_H - MT - MB   # plot height (pixels)

    def w2c(x_m: float, y_m: float) -> tuple[int, int]:
        """World metres → canvas pixel (clamped)."""
        px = int(ML + x_m / x_max * pw)
        py = int(CANVAS_H - MB - y_m / y_max * ph)
        return (max(0, min(CANVAS_W - 1, px)),
                max(0, min(CANVAS_H - 1, py)))

    # ── Canvas ────────────────────────────────────────────────────────────────
    img = np.full((CANVAS_H, CANVAS_W, 3), _BG, dtype=np.uint8)

    # Floor line
    cv2.line(img, w2c(0, 0), w2c(x_max, 0), _DARK_GREY, 1)

    # ── Y-axis tick lines & labels ────────────────────────────────────────────
    for y_tick in [0.0, hood_h_m, hub_h_m]:
        cx, cy = w2c(0, y_tick)
        cv2.line(img, (ML - 5, cy), (ML, cy), _GREY, 1)
        cv2.putText(img, f"{y_tick:.1f}m", (2, cy + 4),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.36, _GREY, 1)

    # ── X-axis tick labels ────────────────────────────────────────────────────
    for x_tick in [0.0, distance_m * 0.5, distance_m]:
        cx, cy = w2c(x_tick, 0)
        cv2.line(img, (cx, cy), (cx, cy + 4), _GREY, 1)
        cv2.putText(img, f"{x_tick:.1f}m", (cx - 14, cy + 16),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.36, _GREY, 1)

    # ── Robot / shooter box (left) ────────────────────────────────────────────
    ROBOT_W_M = 0.65
    ROBOT_H_M = hood_h_m + 0.03
    cv2.rectangle(img, w2c(0, ROBOT_H_M), w2c(ROBOT_W_M, 0), _GREY, 2)

    # Shooter exit point (green dot)
    shoot_px = w2c(ROBOT_W_M * 0.88, hood_h_m)
    cv2.circle(img, shoot_px, 5, _GREEN, -1)

    # Launch-direction arrow
    arr_end = w2c(ROBOT_W_M * 0.88 + 0.22 * math.cos(angle_rad),
                  hood_h_m        + 0.22 * math.sin(angle_rad))
    cv2.arrowedLine(img, shoot_px, arr_end, _GREEN, 2, tipLength=0.35)

    # ── Hub / target box (right) ──────────────────────────────────────────────
    HUB_W_M    = 0.28
    HUB_OPEN_M = 0.28   # opening height
    hub_tl = w2c(distance_m - HUB_W_M, hub_h_m + HUB_OPEN_M / 2)
    hub_br = w2c(distance_m,            hub_h_m - HUB_OPEN_M / 2)
    cv2.rectangle(img, hub_tl, hub_br, _WHITE, 2)

    # Stand from hub base to floor
    stand_top    = w2c(distance_m - HUB_W_M / 2, hub_h_m - HUB_OPEN_M / 2)
    stand_bottom = w2c(distance_m - HUB_W_M / 2, 0)
    cv2.line(img, stand_top, stand_bottom, _DARK_GREY, 1)

    # ── Trajectory curve ──────────────────────────────────────────────────────
    if len(traj) >= 2:
        pts_canvas = [w2c(p[0], p[1]) for p in traj]
        for i in range(len(pts_canvas) - 1):
            cv2.line(img, pts_canvas[i], pts_canvas[i + 1], _TRAJ_COLOR, 2)

    # ── Fuel ball — yellow dot at hub opening ─────────────────────────────────
    ball_px = w2c(distance_m - HUB_W_M * 0.15, hub_h_m)
    cv2.circle(img, ball_px, 9,  _YELLOW, -1)
    cv2.circle(img, ball_px, 9,  _WHITE,   1)   # thin outline

    # ── Info overlay (top-left of plot area) ─────────────────────────────────
    src = "Pi" if pi_active else "RIO"
    info_lines = [
        (f"Dist:   {distance_m * 3.28084:.1f} ft  ({distance_m / 0.0254:.0f}\")", _TEXT_COLOR),
        (f"Speed:  {v_flywheel_fps:.1f} ft/s  (flywheel)",              _TEXT_COLOR),
        (f"Angle:  {angle_deg:.1f}\u00b0  Hood: {hood_h_m * 39.37:.0f}\" "
         f" Hub: {hub_h_m * 39.37:.0f}\"",                              _TEXT_COLOR),
        (f"Effic:  {efficiency * 100:.0f}%   "
         f"Drag: {drag_coeff:.4f} kg/m",                                _LABEL_COLOR),
        (f"Source: {src}",
         _GREEN if pi_active else _LABEL_COLOR),
    ]
    for idx, (text, color) in enumerate(info_lines):
        cv2.putText(img, text, (ML + 6, MT + 10 + idx * 20),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.45, color, 1, cv2.LINE_AA)

    # ── Title ─────────────────────────────────────────────────────────────────
    cv2.putText(img, "Team 539  \u2014  Shot Trajectory",
                (CANVAS_W // 2 - 120, 20),
                cv2.FONT_HERSHEY_SIMPLEX, 0.52, _LABEL_COLOR, 1, cv2.LINE_AA)

    # ── Encode to JPEG ────────────────────────────────────────────────────────
    ok, buf = cv2.imencode(".jpg", img, [cv2.IMWRITE_JPEG_QUALITY, 85])
    return bytes(buf) if ok else b""


# ── Main entry point ──────────────────────────────────────────────────────────

def run(table: ntcore.NetworkTable) -> None:
    """
    Blocking loop.  Called from main.py in a daemon thread.

    :param table: Pre-initialized UltraShooter NT table.
                  The ShooterTuner table is accessed via the same NT instance.
    """
    global _latest_jpeg

    inst         = table.getInstance()
    tuner_params = inst.getTable("ShooterTuner").getSubTable("Params")

    dist_sub     = table.getDoubleTopic("Avg Distance to Hub (ft)").subscribe(0.0)
    target_sub   = table.getDoubleTopic("Target ft/s").subscribe(0.0)
    physics_sub  = table.getDoubleTopic("Physics Velocity ft/s").subscribe(0.0)
    angle_sub    = table.getDoubleTopic("Launch Angle (deg)").subscribe(75.0)
    hood_sub     = table.getDoubleTopic("Hood Height From Floor (in)").subscribe(27.0)
    hub_sub      = table.getDoubleTopic("Hub Center Height From Floor (in)").subscribe(96.0)
    pi_act_sub   = table.getBooleanTopic("Pi Active").subscribe(False)

    effic_sub    = tuner_params.getDoubleTopic("FlywheelEfficiency").subscribe(0.85)
    drag_sub     = tuner_params.getDoubleTopic("DragCoefficient").subscribe(0.0132)
    mass_sub     = tuner_params.getDoubleTopic("BallMassLbs").subscribe(0.595)

    # Start HTTP server in a background thread
    threading.Thread(target=_start_server, daemon=True,
                     name="traj-http").start()
    print(f"[TrajectoryViz] MJPEG stream on port {STREAM_PORT}  "
          f"→ http://10.5.39.XX:{STREAM_PORT}/stream.mjpg")

    while True:
        d       = dist_sub.get() / 3.28084   # ft → m for physics/canvas
        v_fps   = target_sub.get()
        angle   = angle_sub.get()
        hood_h  = hood_sub.get() * 0.0254    # inches → metres
        hub_h   = hub_sub.get()  * 0.0254
        pi_act  = pi_act_sub.get()
        effic   = effic_sub.get()
        drag    = drag_sub.get()
        mass    = mass_sub.get()

        # Use physics velocity if no active setpoint
        if v_fps < 1.0:
            v_fps = physics_sub.get()

        frame = _draw_frame(d, v_fps, angle, hood_h, hub_h,
                            effic, drag, mass, pi_act)
        with _frame_lock:
            _latest_jpeg = frame

        time.sleep(1.0 / FPS)

File: pi-java/src/main/java/TrajectoryVisualizer.java
/**
 * Shot Trajectory Visualizer — Team 539
 *
 * <p>Mirrors pi/trajectory_visualizer.py. Renders a 2D side-view of the current
 * physics-calculated shot as JPEG frames and streams them as MJPEG on port 1184.
 * Uses Java AWT (no OpenCV) so no native dependencies are needed on the Pi.
 *
 * <p>Stream URL: {@code http://10.5.39.XX:1184/stream.mjpg}
 * Add as a camera widget in Elastic.
 *
 * <p>NT reads from "UltraShooter" and "ShooterTuner/Params".
 * Runs at 10 Hz — the trajectory is static, so high frame rate is unnecessary.
 */
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import edu.wpi.first.networktables.BooleanSubscriber;
import edu.wpi.first.networktables.DoubleSubscriber;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

public class TrajectoryVisualizer {

    // ── Stream / canvas settings ──────────────────────────────────────────────────
    private static final int STREAM_PORT = 1184;
    private static final int CANVAS_W    = 800;
    private static final int CANVAS_H    = 420;
    private static final int FPS         = 10;

    // ── Colours ───────────────────────────────────────────────────────────────────
    private static final Color BG          = new Color(28,  28,  28);
    private static final Color WHITE       = new Color(255, 255, 255);
    private static final Color GREY        = new Color(140, 140, 140);
    private static final Color DARK_GREY   = new Color(70,  70,  70);
    private static final Color GREEN       = new Color(80,  200, 80);
    private static final Color YELLOW      = new Color(255, 215, 0);
    private static final Color TRAJ_COLOR  = new Color(210, 210, 210);
    private static final Color TEXT_COLOR  = new Color(225, 225, 225);
    private static final Color LABEL_COLOR = new Color(180, 200, 255);

    // ── Shared JPEG frame buffer ──────────────────────────────────────────────────
    private static final ReentrantLock frameLock   = new ReentrantLock();
    private static volatile byte[]     latestJpeg  = new byte[0];

    // ── Physics helpers ───────────────────────────────────────────────────────────

    /** Returns list of (x, y) world-space points in metres, y from the floor. */
    private static List<double[]> simulateTrajectory(
            double v0Mps, double angleRad, double dragPerMass, double hoodHeightM) {

        List<double[]> pts = new ArrayList<>();
        double vx = v0Mps * Math.cos(angleRad);
        double vy = v0Mps * Math.sin(angleRad);
        double x  = 0, y = hoodHeightM; // start at hood exit height above floor

        pts.add(new double[]{x, y});
        final double dt = 0.005;

        for (int i = 0; i < 4000; i++) {
            double speed = Math.sqrt(vx * vx + vy * vy);
            vx += (-dragPerMass * speed * vx) * dt;
            vy += (-9.81 - dragPerMass * speed * vy) * dt;
            x  += vx * dt;
            y  += vy * dt;
            pts.add(new double[]{x, y});
            if (y < -0.1 || x > 12.0) break;
        }
        return pts;
    }

    // ── Rendering ─────────────────────────────────────────────────────────────────

    private static byte[] drawFrame(
            double distanceM, double vFlywheelFps, double angleDeg,
            double hoodHM, double hubHM,
            double efficiency, double dragCoeff, double ballMassLbs,
            boolean piActive) {

        double angleRad    = Math.toRadians(angleDeg);
        double ballMassKg  = ballMassLbs * 0.453592;
        double dragPerMass = (dragCoeff > 0) ? dragCoeff / Math.max(ballMassKg, 0.001) : 0.0;
        double v0BallMps   = (vFlywheelFps / 3.28084) * efficiency;

        // Simulate trajectory.
        List<double[]> traj = new ArrayList<>();
        if (distanceM > 0.1 && v0BallMps > 0.5) {
            traj = simulateTrajectory(v0BallMps, angleRad, dragPerMass, hoodHM);
        }

        // World bounding box. Guard against zero to prevent divide-by-zero in w2c.
        double xMax = Math.max(distanceM * 1.12, 1.5);
        double yMax = Math.max(hubHM * 1.45, 0.1);
        if (!traj.isEmpty()) {
            for (double[] p : traj) yMax = Math.max(yMax, p[1] * 1.18);
        }

        // Layout margins (pixels).
        final int ML = 58, MR = 18, MT = 32, MB = 38;
        final int pw = CANVAS_W - ML - MR;
        final int ph = CANVAS_H - MT - MB;

        BufferedImage img = new BufferedImage(CANVAS_W, CANVAS_H, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = img.createGraphics();
        g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);

        // World → canvas coordinate helper (captured as lambdas via helper method).
        final double _xMax = xMax, _yMax = yMax;
        java.util.function.BiFunction<Double, Double, int[]> w2c = (wx, wy) -> {
            int px = (int) (ML + wx / _xMax * pw);
            int py = (int) (CANVAS_H - MB - wy / _yMax * ph);
            return new int[]{
                Math.max(0, Math.min(CANVAS_W - 1, px)),
                Math.max(0, Math.min(CANVAS_H - 1, py))
            };
        };

        // Background.
        g.setColor(BG);
        g.fillRect(0, 0, CANVAS_W, CANVAS_H);

        // Floor line.
        int[] fl = w2c.apply(0.0, 0.0), fr = w2c.apply(xMax, 0.0);
        g.setColor(DARK_GREY);
        g.setStroke(new BasicStroke(1));
        g.drawLine(fl[0], fl[1], fr[0], fr[1]);

        // Y-axis ticks.
        g.setFont(new Font("Monospaced", Font.PLAIN, 11));
        for (double yTick : new double[]{0.0, hoodHM, hubHM}) {
            int[] tc = w2c.apply(0.0, yTick);
            g.setColor(GREY);
            g.drawLine(ML - 5, tc[1], ML, tc[1]);
            g.drawString(String.format("%.1fm", yTick), 2, tc[1] + 4);
        }

        // X-axis ticks.
        for (double xTick : new double[]{0.0, distanceM * 0.5, distanceM}) {
            int[] tc = w2c.apply(xTick, 0.0);
            g.setColor(GREY);
            g.drawLine(tc[0], tc[1], tc[0], tc[1] + 4);
            g.drawString(String.format("%.1fm", xTick), tc[0] - 14, tc[1] + 16);
        }

        // Robot / shooter box (left).
        double robotWM = 0.65, robotHM = hoodHM + 0.03;
        int[] rtl = w2c.apply(0.0, robotHM), rbr = w2c.apply(robotWM, 0.0);
        g.setColor(GREY);
        g.setStroke(new BasicStroke(2));
        g.drawRect(rtl[0], rtl[1], rbr[0] - rtl[0], rbr[1] - rtl[1]);

        // Shooter exit dot (green).
        int[] shootPx = w2c.apply(robotWM * 0.88, hoodHM);
        g.setColor(GREEN);
        g.fillOval(shootPx[0] - 5, shootPx[1] - 5, 10, 10);

        // Launch direction arrow.
        int[] arrowEnd = w2c.apply(
            robotWM * 0.88 + 0.22 * Math.cos(angleRad),
            hoodHM         + 0.22 * Math.sin(angleRad));
        g.setColor(GREEN);
        g.setStroke(new BasicStroke(2));
        drawArrow(g, shootPx[0], shootPx[1], arrowEnd[0], arrowEnd[1]);

        // Hub / target box (right).
        double hubWM = 0.28, hubOpenM = 0.28;
        int[] htl = w2c.apply(distanceM - hubWM, hubHM + hubOpenM / 2.0);
        int[] hbr = w2c.apply(distanceM,          hubHM - hubOpenM / 2.0);
        g.setColor(WHITE);
        g.setStroke(new BasicStroke(2));
        g.drawRect(htl[0], htl[1], hbr[0] - htl[0], hbr[1] - htl[1]);

        // Hub stand.
        int[] standTop = w2c.apply(distanceM - hubWM / 2.0, hubHM - hubOpenM / 2.0);
        int[] standBot = w2c.apply(distanceM - hubWM / 2.0, 0.0);
        g.setColor(DARK_GREY);
        g.setStroke(new BasicStroke(1));
        g.drawLine(standTop[0], standTop[1], standBot[0], standBot[1]);

        // Trajectory curve.
        if (traj.size() >= 2) {
            g.setColor(TRAJ_COLOR);
            g.setStroke(new BasicStroke(2));
            for (int i = 0; i < traj.size() - 1; i++) {
                int[] a = w2c.apply(traj.get(i)[0], traj.get(i)[1]);
                int[] b = w2c.apply(traj.get(i + 1)[0], traj.get(i + 1)[1]);
                g.drawLine(a[0], a[1], b[0], b[1]);
            }
        }

        // Ball — yellow dot at hub opening.
        int[] ballPx = w2c.apply(distanceM - hubWM * 0.15, hubHM);
        g.setColor(YELLOW);
        g.fillOval(ballPx[0] - 9, ballPx[1] - 9, 18, 18);
        g.setColor(WHITE);
        g.setStroke(new BasicStroke(1));
        g.drawOval(ballPx[0] - 9, ballPx[1] - 9, 18, 18);

        // Info overlay.
        g.setFont(new Font("Monospaced", Font.PLAIN, 12));
        String src = piActive ? "Pi" : "RIO";
        String[] lines = {
            String.format("Dist:   %.1f ft  (%.0f\")", distanceM * 3.28084, distanceM / 0.0254),
            String.format("Speed:  %.1f ft/s  (flywheel)", vFlywheelFps),
            String.format("Angle:  %.1f\u00b0  Hood: %.0f\"  Hub: %.0f\"",
                          angleDeg, hoodHM * 39.37, hubHM * 39.37),
            String.format("Effic:  %.0f%%   Drag: %.4f kg/m",
                          efficiency * 100, dragCoeff),
            "Source: " + src,
        };
        Color[] lineColors = {TEXT_COLOR, TEXT_COLOR, TEXT_COLOR, LABEL_COLOR,
                              piActive ? GREEN : LABEL_COLOR};
        for (int i = 0; i < lines.length; i++) {
            g.setColor(lineColors[i]);
            g.drawString(lines[i], ML + 6, MT + 10 + i * 20);
        }

        // Title.
        g.setColor(LABEL_COLOR);
        g.setFont(new Font("SansSerif", Font.PLAIN, 13));
        g.drawString("Team 539 \u2014 Shot Trajectory", CANVAS_W / 2 - 100, 20);

        g.dispose();

        // Encode to JPEG.
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
            ImageIO.write(img, "jpg", baos);
            return baos.toByteArray();
        } catch (IOException e) {
            return new byte[0];
        }
    }

    /** Draws a simple arrow from (x1,y1) to (x2,y2). */
    private static void drawArrow(Graphics2D g, int x1, int y1, int x2, int y2) {
        g.drawLine(x1, y1, x2, y2);
        double angle = Math.atan2(y2 - y1, x2 - x1);
        int tipLen = 8;
        double spread = Math.PI / 6;
        int ax1 = (int) (x2 - tipLen * Math.cos(angle - spread));
        int ay1 = (int) (y2 - tipLen * Math.sin(angle - spread));
        int ax2 = (int) (x2 - tipLen * Math.cos(angle + spread));
        int ay2 = (int) (y2 - tipLen * Math.sin(angle + spread));
        g.drawLine(x2, y2, ax1, ay1);
        g.drawLine(x2, y2, ax2, ay2);
    }

    // ── MJPEG HTTP server ─────────────────────────────────────────────────────────

    private static void startServer() {
        try {
            HttpServer server = HttpServer.create(new InetSocketAddress(STREAM_PORT), 0);
            server.createContext("/stream.mjpg", TrajectoryVisualizer::handleStream);
            server.setExecutor(java.util.concurrent.Executors.newCachedThreadPool());
            server.start();
            System.out.printf("[TrajectoryViz] MJPEG stream on port %d → "
                + "http://10.5.39.XX:%d/stream.mjpg%n", STREAM_PORT, STREAM_PORT);
        } catch (IOException e) {
            System.out.println("[TrajectoryViz] Failed to start HTTP server: " + e.getMessage());
        }
    }

    private static void handleStream(HttpExchange exchange) {
        try {
            exchange.getResponseHeaders().set("Content-Type",
                "multipart/x-mixed-replace; boundary=frame");
            exchange.getResponseHeaders().set("Cache-Control", "no-cache");
            exchange.sendResponseHeaders(200, 0);
            OutputStream out = exchange.getResponseBody();

            long frameDelayMs = 1000L / FPS;
            while (true) {
                byte[] jpeg;
                frameLock.lock();
                try { jpeg = latestJpeg; } finally { frameLock.unlock(); }

                if (jpeg.length > 0) {
                    String header = "--frame\r\nContent-Type: image/jpeg\r\n"
                        + "Content-Length: " + jpeg.length + "\r\n\r\n";
                    out.write(header.getBytes());
                    out.write(jpeg);
                    out.write("\r\n".getBytes());
                    out.flush();
                }
                Thread.sleep(frameDelayMs);
            }
        } catch (Exception e) {
            // Client disconnected — normal.
        }
    }

    // ── NT loop ───────────────────────────────────────────────────────────────────

    /**
     * Blocking 10 Hz loop. Called from {@link Main} in a daemon thread.
     *
     * @param table pre-initialized "UltraShooter" NT table
     */
    public static void run(NetworkTable table) {
        NetworkTableInstance inst  = table.getInstance();
        NetworkTable tunerParams   = inst.getTable("ShooterTuner").getSubTable("Params");

        DoubleSubscriber  distSub    = table.getDoubleTopic("Avg Distance to Hub (ft)").subscribe(0.0);
        DoubleSubscriber  targetSub  = table.getDoubleTopic("Target ft/s").subscribe(0.0);
        DoubleSubscriber  physicsSub = table.getDoubleTopic("Physics Velocity ft/s").subscribe(0.0);
        DoubleSubscriber  angleSub   = table.getDoubleTopic("Launch Angle (deg)").subscribe(75.0);
        DoubleSubscriber  hoodSub    = table.getDoubleTopic("Hood Height From Floor (in)").subscribe(27.0);
        DoubleSubscriber  hubSub     = table.getDoubleTopic("Hub Center Height From Floor (in)").subscribe(96.0);
        BooleanSubscriber piActSub   = table.getBooleanTopic("Pi Active").subscribe(false);

        DoubleSubscriber efficSub = tunerParams.getDoubleTopic("FlywheelEfficiency").subscribe(0.85);
        DoubleSubscriber dragSub  = tunerParams.getDoubleTopic("DragCoefficient").subscribe(0.0132);
        DoubleSubscriber massSub  = tunerParams.getDoubleTopic("BallMassLbs").subscribe(0.595);

        // Start the MJPEG HTTP server in a background daemon thread.
        Thread serverThread = new Thread(TrajectoryVisualizer::startServer, "traj-http");
        serverThread.setDaemon(true);
        serverThread.start();

        long frameDelayMs = 1000L / FPS;

        while (true) {
            long start = System.currentTimeMillis();

            double d      = distSub.get() / 3.28084;   // ft → m
            double vFps   = targetSub.get();
            double angle  = angleSub.get();
            double hoodH  = hoodSub.get() * 0.0254;    // in → m
            double hubH   = hubSub.get()  * 0.0254;
            boolean piAct = piActSub.get();
            double effic  = efficSub.get();
            double drag   = dragSub.get();
            double mass   = massSub.get();

            // Fall back to physics velocity if no active setpoint.
            if (vFps < 1.0) vFps = physicsSub.get();

            byte[] frame = drawFrame(d, vFps, angle, hoodH, hubH, effic, drag, mass, piAct);
            frameLock.lock();
            try { latestJpeg = frame; } finally { frameLock.unlock(); }

            long elapsed = System.currentTimeMillis() - start;
            long sleep   = frameDelayMs - elapsed;
            if (sleep > 0) {
                try { Thread.sleep(sleep); }
                catch (InterruptedException e) { Thread.currentThread().interrupt(); return; }
            }
        }
    }
}
